pkg/core/blockchain.go

```diff
@@ -318,26 +323,46 @@ func (bc *Blockchain) AddBlock(block *block.Block) error {
 			}
 		}
 	}
-	headerLen := bc.headerListLen()
-	if int(block.Index) == headerLen {
-		err := bc.AddHeaders(block.Header())
-		if err != nil {
-			return err
-		}
-	}
 	return bc.storeBlock(block)
 }
 
 // AddHeaders processes the given headers and add them to the
 // HeaderHashList.
-func (bc *Blockchain) AddHeaders(headers ...*block.Header) (err error) {
+func (bc *Blockchain) AddHeaders(headers ...*block.Header) error {
+	return bc.addHeaders(bc.config.VerifyBlocks, headers...)
+}
+
+func (bc *Blockchain) addHeaders(verify bool, headers ...*block.Header) (err error) {
 	var (
 		start = time.Now()
 		batch = bc.dao.store.Batch()
 	)
 
+	if len(headers) == 0 {
+		return nil
+	} else if verify {
+		// Verify that the chain of the headers is consistent.
+		// Verification of the first header is performed separately.
+		lastHeader := headers[0]
```

Comment:
Why can't we make a `GetHeader()` here for `headers[0].PrevHash` and iterate whole `headers` range down below?

Reply:
Done!